CFD-384 : Group management Categories : Overview screen.

nader77 opened the pull request:
#571

nader77 commented:
Update: Added all the needed HTML element and styling: (No functionality).
![selection_089](https://cloud.githubusercontent.com/assets/7369740/6996797/38d04458-dba7-11e4-99eb-96ac3ef352d0.png)

nader77 commented:
This is the master branch for the ``Group management Categories``, All branches regarding this issue will be merged here, The theming is finished in this PR.

nader77 commented:
@MPParsley PR is ready for your review.

zero2one commented:
# Path to the categories overview

There is no need to put the vocab ID in the path to the categories management: 1 group can have only 1 categories taxonomy.

So refactor the menu configuration:

```
$items['manage/categories/%taxonomy_vocabulary_machine_name'] = …
```

TO

```
$items['manage/categories'] = …
```

And change the callback so it loads the categories taxonomy based on the context:

```
$group = c4m_og_current_group();
$vocabulary = c4m_og_vocab_load_group_vocabulary($group, 'categories');
```

zero2one commented:
@nader77 needs changes & re-roll

nader77 commented:
@zero2one Fixed the path and the term form.

Category management overview path:
![selection_100](https://cloud.githubusercontent.com/assets/7369740/7218840/463d6578-e68b-11e4-9e62-f81e011221c2.png)

Term form has only the name field:
![selection_099](https://cloud.githubusercontent.com/assets/7369740/7218841/4f1f3dec-e68b-11e4-9eef-d0d808da2a8b.png)

nader77 commented:
PR is ready for code review.

zero2one commented:
@nader77 I got this question while installing capacity4more. What is the proper Angular version we should use and can we fix this so no question pops up (will break Travis-CI builds).

![c4m-what-bower-version](https://cloud.githubusercontent.com/assets/133124/7224911/8e44af90-e738-11e4-825c-b6c3eff046d9.png)

Same for angular-animate
![screen shot 2015-04-20 at 08 39 28](https://cloud.githubusercontent.com/assets/133124/7224927/d5b86aba-e738-11e4-8fdb-47db08b73e33.png)

angular-touch
![screen shot 2015-04-20 at 08 39 36](https://cloud.githubusercontent.com/assets/133124/7224929/ddeff496-e738-11e4-925b-041f3d4d4be6.png)

nader77 commented:
@zero2one This isn't happening in my local install nor in the Travis install, Anyway I've added a resolution to the bower.json file, This should stop it from happening and always picking the right version. ``1.2.16``

zero2one commented:
It seemed to be a problem because I was running the `./scripts/reset -dyl` script.

A complete install fixes the issue (`./install -dyl`).

nader77 commented:
@zero2one Want me to try to fix the ``reset`` script ?

zero2one commented:
@nader77 no, is not needed. The reason why it was not working was because of the big-merge you did. Helena updates last week the AngularJs version (and of other depending libraries). That broke the reset (who does not re-download all).

zero2one commented:
We are missing a button to add a new Category type (level 1 term).

See current platform:
![c4m-add-new-category-type](https://cloud.githubusercontent.com/assets/133124/7226346/66264e96-e748-11e4-99d3-2a1c5ddd2a53.png)

nader77 commented:
@zero2one Sorry this wasn't in the wireframe, Should it link to the 1st level terms overview page ?
Or create term form ?

zero2one commented:
@nader77 The "Add new type of category" should link to the 1st level creation form.

zero2one commented:
Added the improvements for the custom Term forms to a new task: https://issuetracker.amplexor.com/jira/browse/CFD-434

zero2one commented:
I was reviewing the technical implementation of the Category management pages and I have some big concerns how it has been developed (complexity, maintainability).

We really need to avoid big functions of over 40 lines of code that do to much. It's better to split certain functionality in smaller functions.

We really need to refactor this based on the Drupal standards & best practices:

# Manage Category access callback

Add a **c4m_features_og_manage_categories_access()** callback that will be used to validate if the user has access to the category management pages. The user should have access if:

* He accesses the pages within a group context (OG context exists trough PURL prefix).
* The user has access to manage the group categories (use og_access & og_vocab for this).


# Categories

## Categories overview page

Overview of all Categories and their Terms. 
Load the vocabulary using:

```
$group = c4m_og_current_group();
$vocabulary = c4m_og_vocab_load_group_vocabulary($group, 'c4m_vocab_category');
```

* Path: **<group path>/manage/categories**
* page callback: **drupal_get_form**
* page arguments: **c4m_features_og_manage_categories_overview_form**
* access callback: **c4m_features_og_manage_categories_access**

Create a form
* c4m_features_og_manage_categories_overview_form() that returns the fields needed to render the page form.
* Show all Level 2 terms grouped by their parent term.
* Add for each Category type an input field to add a new term. Use the validate & submit handlers to validate & store the data.
* The order of the terms can be changed per Category, use the validate & submit handlers to store the new order.

Add a 
* validate handler: **c4m_features_og_manage_categories_overview_form_validate()**
* submit handler : **c4m_features_og_manage_categories_overview_form_submit()**
to validate and save the data.

## Edit Category term page

A form page to edit an existing Category term.

* Path: <group path>/manage/categories/term/%/edit
* page callback: **drupal_get_form**
* page arguments: **c4m_features_og_manage_categories_term_form**, array(3)
* access callback: **c4m_features_og_manage_categories_access**

Create a form
* **c4m_features_og_manage_categories_term_form()** that returns the fields needed to edit a term:
  * Name
  * Category (parent term), this allows the user to move a term from one category to another. 
* The form can only be used for level 2 terms.
* Validate if the given term id belongs to the current group.

Add a 
* validate handler: **c4m_features_og_manage_categories_term_form_validate()**
* submit handler : **c4m_features_og_manage_categories_term_form_submit()**
to validate and save the data.

## Delete Category term page

A form to confirm the deletion of a Category term.

* Path: <group path>/manage/categories/term/%/delete
* page callback: **drupal_get_form**
* page arguments: **c4m_features_og_manage_categories_term_delete_confirm_form**, array(3)
* access callback: **c4m_features_og_manage_categories_access**

Create a form
* **c4m_features_og_manage_categories_type_delete_confirm_form()** that returns the confirm form using the [confirm_form()](http://dgo.to/a/confirm_form) callback.
* The form can only be used for level 2 terms.
* Validate if the given term id belongs to the current group.

Add a 
* submit handler : **c4m_features_og_manage_categories_term_delete_confirm_form_submit()**
to delete the term.


# Category types

## Category types overview page

Overview of all Category types.

* Path: <group path>/manage/categories/types
* page callback: **drupal_get_form**
* page arguments: **c4m_features_og_manage_categories_types_overview_form**
* access callback: **c4m_features_og_manage_categories_access**

Create a form
* **c4m_features_og_manage_categories_types_overview_form()** that returns the fields needed to render the page form. 
* The form shows only the level 1 terms. 
* The terms order can be changed using drag-n-drop.
* Add an input field to add a new term, use the validate & submit handler to validate and save the new term.

Add a 
* validate handler: **c4m_features_og_manage_categories_types_overview_form_validate()**
* submit handler : **c4m_features_og_manage_categories_types_overview_form_submit()**
to validate and save the data.


## Edit Category type page

A form page to edit an existing Category type.

* Path: <group path>/manage/categories/type/%/edit
* page callback: **drupal_get_form**
* page arguments: **c4m_features_og_manage_categories_type_form**, array(3)
* access callback: **c4m_features_og_manage_categories_access**

Create a form
* **c4m_features_og_manage_categories_type_form()** that returns the fields needed to edit a term:
  * Name
* The form can only be used for level 1 terms.
* Validate if the given term id belongs to the current group.

Add a 
* validate handler: **c4m_features_og_manage_categories_type_form_validate()**
* submit handler : **c4m_features_og_manage_categories_type_form_submit()**
to validate and save the data.


## Delete Category type page

A form to confirm the deletion of a Category type.

* Path: <group path>/manage/categories/type/%/delete
* page callback: **drupal_get_form**
* page arguments: **c4m_features_og_manage_categories_type_delete_confirm_form**, array(3)
* access callback: **c4m_features_og_manage_categories_access**

Create a form
* **c4m_features_og_manage_categories_type_delete_confirm_form()** that returns the confirm form using the [confirm_form()](http://dgo.to/a/confirm_form) callback.
* The form can only be used for level 1 terms.
* Validate if the given term id belongs to the current group.
* Add a warning that deleting the Category will also delete all terms belonging to it.

Add a 
* submit handler : **c4m_features_og_manage_categories_type_form_submit()**
to delete the Category Type.

Make sure to delete all Category terms (level 2) when the type is deleted.

nader77 commented:
Update: Removed paging code.

nader77 commented:
Update: Added page callbacks, Page arguments.

nader77 commented:
Questions:
* Why do we need the **c4m_features_og_manage_categories_types_overview_form_validate()** ? It's a form without any input (Only re-ordering).

* Do we need another page argument for "Category types overview page", We used the same one with some one-liner ``if``s, Saves creating a duplicated code (big one) just for sake of very little changes.

nader77 commented:
Update: Added category term edit form:
![selection_108](https://cloud.githubusercontent.com/assets/7369740/7337106/f267fd88-ec24-11e4-9497-db610aa0f4ae.png)

nader77 commented:
Update: Added a category type custom edit form:
![selection_109](https://cloud.githubusercontent.com/assets/7369740/7337514/34690578-ec35-11e4-8672-bb640bc73996.png)

nader77 commented:
@zero2one @MPParsley 
Questions:
* Why do we need the **c4m_features_og_manage_categories_types_overview_form_validate()** ? It's a form without any input (Only reordering, Add term form validates through RESTful).

* Do we need another page argument for "Category types overview page", We used the same one with some one-liner ``if``s, Saves creating a duplicated code (big one) just for sake of very little changes.

* `` Type of category`` radio buttons should be on the 2nd level terms right ? (The picture is attached to 1st level terms)
Should it be available only for platform admins or non-platform admins as well ?
![23c6f090-e834-11e4-9a23-0558072965af](https://cloud.githubusercontent.com/assets/7369740/7336973/21588c6c-ec1f-11e4-984f-e8018c173b5a.png)

zero2one commented:
@nader77 

### 1. Agree
No need for that.

### 2. Yes we need different callbacks.
* The screens are 2 complete different items.
* The current callback is way to complex and very difficult to maintain. 
* Please put the reorder & delete functionality in separate menu paths & callbacks. This is how Drupal is meant to be used and should be used.

We are using the Drupal "MVC" here where the menu items are feeding the controller with different actions. Each action should be an entry in the hook_menu() and each of them should have a specific callback to handle the action.

You will see that the code is much simpeler to read and to maintain if you create seperate callbacks and put all common code in separate helper functions.


### 3. Correct, it's only for the second level (the radio buttons are the first level).
This is available for anybody who can manage the group categories:
  * Platform administrators
  * Group owners
  * Group administrators

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+
+  $parameters = drupal_get_query_parameters();
+  // Check for confirmation forms.
+  if (isset($parameters['delete'])) {
```
Make separate menu entry & callback.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+    // Delete confirmation form.
+    return array_merge($form, c4m_features_og_manage_taxonomy_term_confirm_delete($form, $form_state, $parameters['delete']));
+  }
+  if (isset($parameters['reset_alphabetical'])) {
```
Make separate menu entry & callback.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+  $form['#tree'] = TRUE;
+  $form['#parent_fields'] = FALSE;
+
+  $parameters = drupal_get_query_parameters();
```
Don't invent your own menu router, Drupal has one build in (hook_menu).

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+  $menu_item = menu_get_item(current_path());
+  $tree = $menu_item['path'] == 'manage/categories/types' ? taxonomy_get_tree($vocabulary->vid, 0, 1) : taxonomy_get_tree($vocabulary->vid);
+  $term = current($tree);
+  do {
```
Use always `while` loops instead of  `do {} while();`.

```
while($term = next($tree)) {
  …
}
```

Or use foreach:

```
foreach ($tree as $term) {
  …
}
```

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+    $form_state['redirect'] = array($_GET['q'], (isset($_GET['page']) ? array('query' => array('page' => $_GET['page'])) : array()));
+  }
+
+  if ($menu_item['path'] == 'manage/categories') {
```
No need for this if we split the callback in 2

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+    t('Are you sure you want to delete the term %title?',
+      array('%title' => $term->name)),
+    'manage/categories/' . $path,
+    '<div class="question">'
```
Don't put HTML in here.
Why do you repeat the question again in the description?

You can add extra form elements to the $form array (like markup) to add extra help and fields to the form.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+  taxonomy_check_vocabulary_hierarchy($form['#vocabulary'], $form_state['values']);
+  drupal_set_message(t('Deleted term %name.', array('%name' => $form_state['values']['name'])));
+  watchdog('taxonomy', 'Deleted term %name.', array('%name' => $form_state['values']['name']), WATCHDOG_NOTICE);
+  $menu_item = menu_get_item(current_path());
```
Use 2 different form callbacks to delete the term and type.
Each of them can have their own redirect path.
Set the redirect path in the form declaration.

Add a helper callback to delete the actual term.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+      t('Reset to alphabetical'),
+      t('Cancel'));
+  }
+  else {
```
Use a separate callback.
Put the common functionality in a helper function.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+    $condition = ' AND h.parent = :parent_tid';
+  }
+
+  db_query(
```
Use the [db_update](https://api.drupal.org/api/drupal/includes!database!database.inc/function/db_update/7) function to get an [UpdateQuery](https://api.drupal.org/api/drupal/includes%21database%21query.inc/class/UpdateQuery/7) object. Use that object to dynamically create the update query and to execute it.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+ * @see taxonomy_overview_terms_submit()
+ */
+function c4m_features_og_manage_categories_overview_form_submit($form, &$form_state) {
+  if ($form_state['triggering_element']['#value'] == t('Delete')) {
```
Put in separate callback + submit handler.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+function c4m_features_og_manage_categories_overview_form_submit($form, &$form_state) {
+  if ($form_state['triggering_element']['#value'] == t('Delete')) {
+    // Execute the term deletion.
+    if ($form_state['values']['delete'] === TRUE) {
```
Put in separate callback + submit handler.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+    }
+  }
+
+  if ($form_state['triggering_element']['#value'] == t('Reset to alphabetical')) {
```
Put in separate callback + submit handler.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+  $changed_terms = array();
+
+  $menu_item = menu_get_item(current_path());
+  $tree = $menu_item['path'] == 'manage/categories/types' ? taxonomy_get_tree($vocabulary->vid, 0, 1) : taxonomy_get_tree($vocabulary->vid);
```
Put in separate callback + submit handler.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+ * @ingroup forms
+ * @see taxonomy_form_term()
+ */
+function c4m_features_og_manage_categories_term_form($form, &$form_state, $tid) {
```
This function is handling to much. 1 function = 1 functionality. Keep the "controllers" simple.

Use separate paths, callbacks & submit handlers to keep the code clean. Put common functionality in separate helper method.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+    $form_state['confirm_delete'] = TRUE;
+    return;
+  }
+  if ($form_state['triggering_element']['#value'] == t('Cancel')) {
```
No need for this if you use the [confirm_form](https://api.drupal.org/api/drupal/modules%21system%21system.module/function/confirm_form/7).

Parameters

* $form: Additional elements to add to the form. These can be regular form elements, #value elements, etc., and their values will be available to the submit handler.
* $question: The question to ask the user (e.g. "Are you sure you want to delete the block <em>foo</em>?"). The page title will be set to this value.
* **$path: The page to go to if the user cancels the action**. This can be either:
  * A string containing a Drupal path.
  * An associative array with a 'path' key. Additional array values are passed as the $options parameter to l().
  * If the 'destination' query parameter is set in the URL when viewing a confirmation form, that value will be used instead of $path.
* $description: Additional text to display. Defaults to t('This action cannot be undone.').
* $yes: A caption for the button that confirms the action (e.g. "Delete", "Replace", ...). Defaults to t('Confirm').
* $no: A caption for the link which cancels the action (e.g. "Cancel"). Defaults to t('Cancel').
* $name: The internal name used to refer to the confirmation item.

zero2one commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,834 @@
+  entity_form_submit_build_entity('taxonomy_term', $term, $form, $form_state);
+
+  // Convert text_format field into values expected by taxonomy_term_save().
+  $description = $form_state['values']['description'];
```
Our form does not have a description field, can we make it empty by default?

zero2one commented on capacity4more/modules/c4m/restful/c4m_restful_quick_post/components/c4m-app/src/controllers/addTerm.js:
```diff
@@ -0,0 +1,60 @@
+/**
```
Don't use AngularJs here. Drupal has build in functionality to handle this type of functionality.

nader77 commented:
@HelenaEksler All the required functionality is done, Please go through the code and check it (Specially [this file](https://github.com/capacity4dev/capacity4more/pull/581/files#diff-d599d7ebec10149d8f75d3d4f06c3e02)).

I have tested all the functionality (Add, Edit, Delete, Reorder) over and over again, But it would be great if you can do it one more time before assigning to Peter, Thanks.

HelenaEksler commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,966 @@
+      }
+    }
+
+    if (empty($rows[$parent_id])) {
```
Couldn't it be just else?

HelenaEksler commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,966 @@
+    );
+    // Add necessary classes to rows.
+    $row_position = 0;
+    if(!empty($rows[$parent_id])) {
```
space before "("

HelenaEksler commented on capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc:
```diff
@@ -0,0 +1,966 @@
+  if (isset($params['destination'])) {
+    $form['#destination'] = $params['destination'];
+  }
+  if(!$term = taxonomy_term_load($tid)) {
```
space before "("

HelenaEksler commented:
@nader77 except few minor fixes looks great. After fixes - assign Peter. Also it's worth to add some tests.

nader77 commented:
Update:
* Added "Add a new type of category" functionality to the button.
* Added a new access callback to check if the current term ID is good and that the term belongs to the current group, Had to separate to 2 access callbacks:
  1. For pages with no term ID, just checks current group and OG access to "administer group".
  2. For pages with term ID (e.g: edit forms) checks current group, valid term ID, Term belongs to the current group and OG access to "administer group".

Tried to keep them in the same access callback since there's some similarity, but it resulted in bugs and some access errors.

Workflow video will follow.

nader77 commented:
Workflow video for Manage page, edit forms, Delete, Instant add term and Add new type form:
![manage-categories-2015-05-03_11](https://cloud.githubusercontent.com/assets/7369740/7444324/a573511a-f185-11e4-832a-f1bf9938035b.gif)

nader77 commented:
@HelenaEksler Please review.

nader77 commented:
Merged main branch.
